### PR TITLE
Revamped CI

### PR DESCRIPTION
**Pull request by capnkenny**
- Removed manual build/generate/cibuild scripts
- Added macOS builds (using macOS 10.15 hosted runners)
- Added a manual publishing workflow
- Combined all scheduled flows into one GH Workflow
- Tweaked a whole bunch of other stuff
- Also tweaked the badges (since we would no longer have split workflows)
- Added Unit Test Publishing and error/warning matching (Thanks @FiniteReality !)

... there's probably more but I'm tired.    : ^ )

### CODE REVIEW

**Comment by capnkenny** [comment]
I'm not surprised if all the old checks fail - they'd need to be bypassed and/or removed after this PR is manually reviewed and green-lit (nothing until then).

**Comment by capnkenny** [comment]
Need to fix doxygen! Please hold

File: .github/workflows/build-system.yml
Comment (capnkenny) [inline .github/workflows/build-system.yml:18]: Switched to `MinSizeRel` and `RelWithDebInfo` to replace `Release` and `Debug`

Doesn't seem to cause issues running, however it appears it will cause longer build times on CI since the Conan packages don't have bins for these configs readily available.

File: .github/workflows/build-system.yml
Comment (capnkenny) [inline .github/workflows/build-system.yml:43]: macOS builds were taking over an hour, so I exclude them and re-include them under normal `Release` and `Debug` configs.

Takes ~5 minutes 😄

File: .github/workflows/build-system.yml
Comment (capnkenny) [inline .github/workflows/build-system.yml:59]: Based these steps off the readme so even CI uses the steps that we provide.

File: .github/workflows/build-system.yml
Comment (capnkenny) [inline .github/workflows/build-system.yml:73]: This will only have formatting run when its not against the main branch (so essentially any PRs or manual workflow requests not against the main branch)

File: .github/workflows/doxygen.yml
Comment (capnkenny) [inline .github/workflows/doxygen.yml:42]: There's no easy way to get Conan to just run target doxy, so I spelled out the full commands here.

File: CMakeLists.txt
Comment (capnkenny) [inline CMakeLists.txt:54]: Just a reminder - this hushes Conan from spewing a lot of verbose information we don't need via CMake

File: .github/workflows/publish_release.yml
Comment (capnkenny) [inline .github/workflows/publish_release.yml:89]: This file is essentially all licences for binaries that are redistributed and require attribution + NovelRT's licence. Its only copied for macOS and Windows due to the third party dependency distribution - so Linux is exempt from needing this IIRC.

File: README.md
Comment (capnkenny) [inline README.md:8]: We lose 4 individual "workflow" badges, so I added two specific ones here that are customisable somewhat - one for the nightly builds, and one for releases.

**Review by capnkenny — COMMENTED** [review]
Doxy is fixed, and I left a few comments to explain bits that might need explaining.

I've left my fork's version of the releases up for now here: https://github.com/capnkenny/NovelRT/releases/tag/v0.0.7
(I'll be removing them once this PR is eventually closed - hopefully lol)

File: .github/workflows/build-system.yml
Comment (FiniteReality) [reply .github/workflows/build-system.yml:18]: That is rather unfortunate. Remind me to look into making Dockerfiles for each configuration so we can just pre-build these.

File: .github/workflows/build-system.yml
Comment (FiniteReality) [reply .github/workflows/build-system.yml:43]: That's a :harold: moment if I've ever heard one. If I do the dockerfile crap this should simplify things a LOT.

File: .github/workflows/publish_release.yml
Comment (FiniteReality) [reply .github/workflows/publish_release.yml:89]: We should probably copy the file on Linux too just for safety's sake.

File: .github/workflows/publish_release.yml
Comment (FiniteReality) [inline .github/workflows/publish_release.yml:17]: If we're using this, we'll have to change our commit message convention to https://www.conventionalcommits.org/en/v1.0.0/ as this is how the tool knows to bump the right version number.

File: .github/workflows/publish_release.yml
Comment (FiniteReality) [inline .github/workflows/publish_release.yml:63]: This is only necessary for the gitversion stuff I have in Fabulist - if we're not using it then we can ignore it

File: README.md
Comment (FiniteReality) [reply README.md:8]: We should probably drop include_prereleases here and only show the latest *stable* release - the release pipeline in Fabulist marks releases as prerelease because it's totally possible that a `main` build could have bugs in it, and it's intended that after some human intervention we will manually mark releases as stable.

File: .github/workflows/publish_release.yml
Comment (FiniteReality) [reply .github/workflows/publish_release.yml:17]: Alternatively: we could only run this pipeline when we manually push version tags:
```yml
on:
  create:
    tags:
    - v*
```
That way it will automatically pull the correct version information in the create_release step, and we won't have to figure that out from the commit history. We *will* however have to generate a changelog (which github seems to do automatically: https://docs.github.com/en/repositories/releasing-projects-on-github/automatically-generated-release-notes)

Comment (capnkenny) [reply .github/workflows/publish_release.yml:17]: Personally, I like the convention based on its readability.

@RubyNova what's your preference here? And you, @FiniteReality ?

File: .github/workflows/publish_release.yml
Comment (capnkenny) [reply .github/workflows/publish_release.yml:89]: Makes sense - I'll switch the Linux one to copying LICENCE-DIST.md instead of just the NovelRT licence

File: .github/workflows/publish_release.yml
Comment (FiniteReality) [reply .github/workflows/publish_release.yml:17]: I'm fine with it too, it's just making sure everything follows it.

Comment (FiniteReality) [reply .github/workflows/publish_release.yml:17]: For example, in PRs we'll likely have to squash or rebase commits in order to ensure that they correctly follow the convention if they don't already.

Comment (capnkenny) [reply .github/workflows/publish_release.yml:17]: After talking with @RubyNova, it sounds like we're good to switch over. I've enabled squashes for the repo, and will probably disable merge commits once this is in.
Going to take a note to add additional documentation on this as part of #426

File: .github/workflows/publish_release.yml
Comment (FiniteReality) [inline .github/workflows/publish_release.yml:167]: I don't know how this managed to work in Fabulist but you also need to pass `-X DELETE --silent` to actually delete the ref, otherwise it'll just do a GET request.

Comment (FiniteReality) [reply .github/workflows/publish_release.yml:167]: I don't know how I managed to miss it either - I checked my bash history and it's correct there:
```
$ cat ~/.bash_history | grep "refs/tags"
gh api "repos/novelrt/fabulist/git/refs/tags/v0.2.1" -X DELETE --silent
gh api "repos/novelrt/fabulist/git/refs/tags/v0.2.0" -X DELETE --silent
```
But somehow I just completely missed them from the CI stuff :upside_down_face:

Comment (capnkenny) [reply .github/workflows/publish_release.yml]: _woops_ lol, just fixed this now. Thank you for the heads up!

**Review by RubyNova — APPROVED** [review]
LGTM assuming it all works!

**Review by FiniteReality — APPROVED** [review]
Let's ship it and see if everything burns!

File: .github/workflows/publish_release.yml
Comment (FiniteReality) [inline .github/workflows/publish_release.yml:70]: Not a major issue worth blocking this PR any more for but this should probably be `Install dependencies` rather than `Generate Build System`.

**Pull request merged**
